allow override of wcs when computing sky region in roi_subset_state_to_region

cshanahan1 opened the pull request:
In JDaviz (Cubeviz, specifically), sometimes subsets don't contain spatial wcs information in subset.xatt.coords, which is what is used for the pixel-to-sky conversion in `roi_subset_state_to_region`. To fix this issue, this PR adds the option to provide a different WCS for the conversion, which will override anything in `.coords`. 

(See https://github.com/spacetelescope/jdaviz/pull/2496)

pllim commented on glue_astronomy/translators/regions.py:
```diff
@@ -53,12 +53,24 @@ def range_to_rect(data, ori, low, high):
 
 
-def roi_subset_state_to_region(subset_state, to_sky=False):
+def roi_subset_state_to_region(subset_state, to_sky=False, override_wcs=None):
```
```suggestion
def roi_subset_state_to_region(subset_state, to_sky=False):
```

cshanahan1 replied:
done

pllim commented on glue_astronomy/translators/regions.py:
```diff
@@ -53,12 +53,24 @@ def range_to_rect(data, ori, low, high):
+
+    Parameters
+    ----------
+    subset_state : `glue.core.data.Data`
+        The 2D glue Data object on which the range subset is defined.
```
This is wrong. Input is the subset state, not Data.

pllim commented on glue_astronomy/translators/regions.py:
```diff
@@ -53,12 +53,24 @@ def range_to_rect(data, ori, low, high):
+    ----------
+    subset_state : `glue.core.data.Data`
+        The 2D glue Data object on which the range subset is defined.
+    to_sky: bool
+        Specifies if the range limits are for the x-axis or y-axis respectively.
+    override_wcs : WCS object
+        A world coordinate system (WCS) transformation that
+        supports the `astropy shared interface for WCS
+        <https://docs.astropy.org/en/stable/wcs/wcsapi.html>`_ (e.g.,
+        `astropy.wcs.WCS`, `gwcs.wcs.WCS`).
```
```suggestion
    to_sky: bool or `astropy.wcs.wcsapi.BaseHighLevelWCS`
        If `False`, pixel region is returned; otherwise sky region.
        If `True`, sky region will be computed using its parent's WCS.
        If a WCS is given, then that would be used instead of its parent's WCS
        (not recommended unless you know what you are doing).
```

cshanahan1 replied:
i did something similar, how does that look?

pllim commented on glue_astronomy/translators/regions.py:
```diff
@@ -93,7 +105,10 @@ def roi_subset_state_to_region(subset_state, to_sky=False):
         raise NotImplementedError(f"ROIs of type {roi.__class__.__name__} are not yet supported")
 
     if to_sky:
```
```suggestion
```

pllim commented on glue_astronomy/translators/regions.py:
```diff
@@ -93,7 +105,10 @@ def roi_subset_state_to_region(subset_state, to_sky=False):
 
     if to_sky:
-        reg = reg.to_sky(subset_state.xatt.parent.coords)
+        if override_wcs is not None:
+            reg = reg.to_sky(override_wcs)
+        else:
+            reg = reg.to_sky(subset_state.xatt.parent.coords)
```
Do this at module-level imports:

```python
from astropy.wcs.wcsapi import BaseHighLevelWCS
```

Then this:

```suggestion
    if to_sky is True:
        reg = reg.to_sky(subset_state.xatt.parent.coords)
    elif isinstance(to_sky, BaseHighLevelWCS):
        reg = reg.to_sky(to_sky)
```

cshanahan1 replied:
done

dhomeier commented on glue_astronomy/translators/regions.py:
```diff
@@ -93,7 +105,10 @@ def roi_subset_state_to_region(subset_state, to_sky=False):
+        if override_wcs is not None:
+            reg = reg.to_sky(override_wcs)
+        else:
+            reg = reg.to_sky(subset_state.xatt.parent.coords)
```
```suggestion
            reg = reg.to_sky(subset_state.xatt.parent.coords)
        elif to_sky:
            warnings.warn(f"Invalid input {to_sky} for ``to_sky``, "
                          "must be a valid WCS or set to `True`.",
                          UserWarning)
```
This is really for the change suggested by @pllim, but I cannot add suggestions on the suggestion...
But if we go that path, I think at that point (`not isinstance(to_sky, BaseHighLevelWCS)`) a sanity check is due, else it will appear to just silently ignore the `to_sky` input.

dhomeier reviewed (changes requested):
See comments on @pllim's comments; I am undecided whether re-using the existing kwarg this way is the best option...

And it would be good to test this, e.g. add to the end of `test_rectangular_roi` something like
```python
        self.data.coords = None
        with pytest.raises(**missing_wcs_error**):
            reg_sky = roi_subset_state_to_region(subset_state, to_sky=True)

        reg_sky = roi_subset_state_to_region(subset_state, to_sky=WCS_CELESTIAL)
        assert isinstance(reg_sky, RectangleSkyRegion)
```
Ideally this should perhaps even use two different WCS to make sure the user-provided is used; maybe one could instead pass in the second attempt a deepcopy of `WCS_CELESTIAL` modified to `wcs.wcs.ctype = ['GLON-CAR', 'GLAT-CAR']` and then check `reg_sky.wcs.ctype`...

cshanahan1 commented:
Looks like I pushed some commits i made before seeing the most recent review comments - I added tests (and moved the old one that tested to_sky) and implemented @pllim of overloading to_sky to allow a new WCS to override anything attached to the subset.

dhomeier commented:
Thanks, tests look good! I still would just like to see the exception for an invalid or missing wcs in `.coords` (the original issue above) tested.

pllim commented:
> Ideally this should perhaps even use two different WCS to make sure the user-provided is used

Let's not do that. Ideally, subset state should always refer to its parent. But in Cubeviz, we have to hack around the fact that cube ingested into Cubeviz somehow contains `.coords` without the original spatial WCS info (I cannot find that issue for cross-linking right now). So ideally when this problem is fixed properly, no one will ever have to pass in `to_sky` as a WCS ever again.

pllim commented on glue_astronomy/translators/regions.py:
```diff
@@ -53,12 +54,24 @@ def range_to_rect(data, ori, low, high):
 
 
-def roi_subset_state_to_region(subset_state, to_sky=False):
+def roi_subset_state_to_region(subset_state, to_sky=False, override_wcs=None):
```
This is now not used anymore.
```suggestion
def roi_subset_state_to_region(subset_state, to_sky=False):
```

pllim commented on glue_astronomy/translators/tests/test_regions.py:
```diff
@@ -49,8 +50,6 @@ def test_rectangular_roi(self):
         assert_allclose(reg.width, 2.5)
         assert_allclose(reg.height, 3.5)
 
-        reg_sky = roi_subset_state_to_region(subset_state, to_sky=True)
-        assert isinstance(reg_sky, RectangleSkyRegion)
```
Why was this removed? This is a completely different test.

cshanahan1 replied:
it was the only place that `roi_subset_state_to_region` was being tested, it seemed to me like it was just thrown into an existing test to make sure it worked. i moved it to its own test (just using circle instead of rectangle. i could test every shape but that seems like overkill)

pllim replied:
This one tests the rectangle. Your tests only have circles. I think it does not hurt to leave this in.

pllim replied:
While I understand your point above, I don't see why extra testing here would hurt. This is not an intensive operation.

dhomeier replied:
I agree the original test seemed thrown in a bit randomly, but if you really think Rectangle should be tested separately, perhaps do it properly indeed and `pytest.parametrize` over all supported ROI types (may indeed make sense, since the pixel_to_world transformations could be somewhat different under the hood).

pllim replied:
Well, as a maintainer, Derek has the final say here, so I guess we can remove it. 🤷

dhomeier replied:
Maybe `parametrize`ing the full `test_to_sky` over all ROI types with matching inputs and expected SkyRegions is a bit overkill indeed, so as a compromise you put that back in and add the same basic check to the tests for Polygonal, Elliptical and Point.

pllim commented:
Actually.... I just realized that you do not really need this PR, @cshanahan1 . Since you only use this function to compute pixel region, and then convert to sky region internally in Jdaviz, given that you want to keep both pixel and sky.

😬

dhomeier commented:
> Let's not do that. Ideally, subset state should always refer to its parent. But in Cubeviz, we have to hack around the fact that cube ingested into Cubeviz somehow contains `.coords` without the original spatial WCS info

Thanks for the background – that explains why it all appears a bit hackish atm :D!
So you think no test for invalid input is needed either?

pllim commented:
I actually don't think this PR is needed at all. I am going to revisit https://github.com/spacetelescope/jdaviz/pull/2496 shortly.

cshanahan1 commented:
I parameterized the tests which is also something we can keep even if we don't move forward with allowing 'to_sky' to be a WCS.

cshanahan1 commented:
Closing this for now, doc + test changes we want to keep are in https://github.com/glue-viz/glue-astronomy/pull/97
